File: ycs-golang/content/content_type.go
package content

import (
	"fmt"
	"ycs/contracts"
)

const ContentTypeRef = 7

// ContentType represents type content
type ContentType struct {
	contentType contracts.IAbstractType
}

// NewContentType creates a new ContentType instance
func NewContentType(contentType contracts.IAbstractType) *ContentType {
	return &ContentType{
		contentType: contentType,
	}
}

// GetRef returns the reference ID for this content type
func (c *ContentType) GetRef() int {
	return ContentTypeRef
}

// GetCountable returns whether this content is countable
func (c *ContentType) GetCountable() bool {
	return false
}

// GetLength returns the length of this content
func (c *ContentType) GetLength() int {
	return 1
}

// GetContent returns the content as an interface slice
func (c *ContentType) GetContent() []interface{} {
	return []interface{}{c.contentType}
}

// Copy creates a copy of this content
func (c *ContentType) Copy() contracts.IContent {
	// ContentType cannot be copied directly
	return nil
}

// Splice splits this content at the given offset
func (c *ContentType) Splice(offset int) contracts.IContent {
	// ContentType cannot be split
	return nil
}

// MergeWith attempts to merge this content with another
func (c *ContentType) MergeWith(right contracts.IContent) bool {
	// ContentType cannot be merged
	return false
}

// Integrate integrates this content into a transaction
func (c *ContentType) Integrate(transaction contracts.ITransaction, item contracts.IStructItem) {
	// Implementation would go here
}

// Delete deletes this content
func (c *ContentType) Delete(transaction contracts.ITransaction) {
	// Implementation would go here
}

// Gc garbage collects this content
func (c *ContentType) Gc(store contracts.IStructStore) {
	// Implementation would go here
}

// Write writes this content to an encoder
func (c *ContentType) Write(encoder contracts.IUpdateEncoder, offset int) error {
	// Implementation would go here
	return nil
}

// Type reader registry for avoiding circular dependencies
var typeReaderRegistry = make(map[uint32]func(contracts.IUpdateDecoder) contracts.IAbstractType)

// RegisterTypeReader registers a type reader for a given type reference ID
func RegisterTypeReader(key uint32, reader func(contracts.IUpdateDecoder) contracts.IAbstractType) {
	typeReaderRegistry[key] = reader
}

// ReadContentType reads ContentType from a decoder
func ReadContentType(decoder contracts.IUpdateDecoder) *ContentType {
	if typeReaderRegistry == nil {
		panic("TypeReaderRegistry not initialized. Call RegisterTypeReader() first.")
	}

	typeRef := decoder.ReadTypeRef()
	if reader, exists := typeReaderRegistry[typeRef]; exists {
		abstractType := reader(decoder)
		return NewContentType(abstractType)
	} else {
		panic(fmt.Sprintf("No type reader registered for typeRef %d", typeRef))
	}
}

File: src/Ycs.Content/ContentType.cs
// ------------------------------------------------------------------------------
//  <copyright company="Microsoft Corporation">
//      Copyright (c) Microsoft Corporation.  All rights reserved.
//  </copyright>
// ------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using Ycs.Contracts;

namespace Ycs.Content
{
    public class ContentType : IContentEx
    {
        internal const int _ref = 7;

        internal ContentType(IAbstractType type)
        {
            Type = type;
        }

        int IContentEx.Ref => _ref;

        public bool Countable => true;
        public int Length => 1;

        public IAbstractType Type { get; }

        public IReadOnlyList<object> GetContent() => new object[] { Type };

        public IContent Copy() => new ContentType(Type.InternalCopy());

        public IContent Splice(int offset) => throw new NotImplementedException();

        public bool MergeWith(IContent right) => false;

        void IContentEx.Integrate(ITransaction transaction, IStructItem item)
        {
            Type.Integrate(transaction.Doc, item);
        }

        void IContentEx.Delete(ITransaction transaction)
        {
            var item = Type.Start;

            while (item != null)
            {
                if (!item.Deleted)
                {
                    item.Delete(transaction);
                }
                else
                {
                    // This will be gc'd later and we want to merge it if possible.
                    // We try to merge all deleted items each transaction,
                    // but we have no knowledge about that this needs to merged
                    // since it is not in transaction. Hence we add it to transaction._mergeStructs.
                    transaction.MergeStructs.Add(item);
                }

                item = item.Right as IStructItem;
            }

            foreach (var valueItem in Type.Map.Values)
            {
                if (!valueItem.Deleted)
                {
                    valueItem.Delete(transaction);
                }
                else
                {
                    // Same as above.
                    transaction.MergeStructs.Add(valueItem);
                }
            }

            transaction.Changed.Remove(Type);
        }

        void IContentEx.Gc(IStructStore store)
        {
            var item = Type.Start;
            while (item != null)
            {
                item.Gc(store, parentGCd: true);
                item = item.Right as IStructItem;
            }

            Type.Start = null;

            foreach (var kvp in Type.Map)
            {
                var valueItem = kvp.Value;
                while (valueItem != null)
                {
                    valueItem.Gc(store, parentGCd: true);
                    valueItem = valueItem.Left as IStructItem;
                }
            }

            Type.Map.Clear();
        }

        void IContentEx.Write(IUpdateEncoder encoder, int offset)
        {
            Type.Write(encoder);
        }


        private static Dictionary<uint, Func<IUpdateDecoder, IAbstractType>> _typeReaderRegistry =
            new Dictionary<uint, Func<IUpdateDecoder, IAbstractType>>();

        internal static void RegisterTypeReader(uint key, Func<IUpdateDecoder, IAbstractType> value)
        {
            _typeReaderRegistry.TryAdd(key, value);
        }

        internal static ContentType Read(IUpdateDecoder decoder)
        {
            if (_typeReaderRegistry == null)
            {
                throw new InvalidOperationException("TypeReaderRegistry not initialized. Call ContentType.SetTypeReaderRegistry() first.");
            }

            var typeRef = decoder.ReadTypeRef();
            if (_typeReaderRegistry.TryGetValue(typeRef, out var func))
            {
                var type = func(decoder);
                return new ContentType(type);
            }
            else
            {
                throw new InvalidOperationException($"No type reader registered for typeRef {typeRef}.");
            }
        }


    }
}
